core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactoryImpl.java

```diff
@@ -651,6 +651,14 @@ public SqlTypeName getSqlTypeName() {
       }
       return typeName;
     }
+
+    @Override public int getPrecision() {
+      if (java.sql.Timestamp.class == clazz) {
+        // Timestamp class can hold fractions up to nano seconds.
+        return 9;
+      }
+      return super.getPrecision();
+    }
```

Comment:
Without this, Timestamp fraction part gets lost when the value is cast to TIMESTAMP (without precision). This affects where clause conditions with '='. E.g. `where \"ts\" = TIMESTAMP '2018-12-14 18:29:34.123'` does not match if the 'ts' has the same timestamp because the fraction part `123` dropped.

Reply:
I wonder what should be the result of `cast(timestamp '2018-12-14 18:29:34.123' as timestamp(2)) = timestamp '2018-12-14 18:29:34.123'`

@ijokarumawak , have you though of that case?

By the way, I appreciate your comment here, however it would be so much better to have that comment as a Java comment in the sources.

Reply:
Good point. I haven't thought casting with a precision..
Now I'm adding unit test cases for that. Ideally casting java.sql.Timestamp should work the same as timestamp literals.

I'm going to add the comment to the source, too.